samples/bluetooth/central_dfu_smp/src/main.c

```diff
@@ -0,0 +1,396 @@
+{
+	printk("MTU exchange %s\n", err == 0 ? "successful" : "failed");
+	printk("Current MTU: %u\n", bt_gatt_get_mtu(conn));
+}
+
+static void connected(struct bt_conn *conn, u8_t conn_err)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+	int err;
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	if (conn_err) {
+		printk("Failed to connect to %s (%u)\n", addr, conn_err);
+		return;
+	}
+
+	printk("Connected: %s\n", addr);
+
+	if (bt_conn_security(conn, BT_SECURITY_MEDIUM)) {
+		printk("Failed to set security\n");
+	}
+
+	exchange_params.func = exchange_func;
+	err = bt_gatt_exchange_mtu(conn, &exchange_params);
+	if (err) {
+		printk("MTU exchange failed (err %d)\n", err);
+	} else {
+		printk("MTU exchange pending\n");
+	}
+
+	if (conn == default_conn) {
+		err = bt_gatt_dm_start(conn,
+				       DFU_SMP_UUID_SERVICE,
+				       &discovery_cb,
+				       NULL);
+		if (err) {
+			printk("Could not start the discovery procedure "
+			       "(err %d)\n", err);
+		}
+	}
+}
+
+static void disconnected(struct bt_conn *conn, u8_t reason)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+	int err;
+
+	bt_addr_le_to_str(bt_conn_get_dst(conn), addr, sizeof(addr));
+
+	printk("Disconnected: %s (reason %u)\n", addr, reason);
+
+	if (default_conn != conn) {
+		return;
+	}
+
+	bt_conn_unref(default_conn);
+	default_conn = NULL;
+
+	/* This demo doesn't require active scan */
+	err = bt_scan_start(BT_SCAN_TYPE_SCAN_ACTIVE);
+	if (err) {
+		printk("Scanning failed to start (err %d)\n", err);
+	}
+}
+
+static struct bt_conn_cb conn_callbacks = {
+	.connected = connected,
+	.disconnected = disconnected,
+};
+
+static void scan_init(void)
+{
+	int err;
+
+	struct bt_scan_init_param scan_init = {
+		.connect_if_match = 1,
+		.scan_param = NULL,
+		.conn_param = BT_LE_CONN_PARAM_DEFAULT
+	};
+
+	bt_scan_init(&scan_init);
+	bt_scan_cb_register(&scan_cb);
+
+	err = bt_scan_filter_add(BT_SCAN_FILTER_TYPE_UUID,
+				 DFU_SMP_UUID_SERVICE);
+	if (err) {
+		printk("Scanning filters cannot be set (err %d)\n", err);
+
+		return;
+	}
+
+	err = bt_scan_filter_enable(BT_SCAN_UUID_FILTER, false);
+	if (err) {
+		printk("Filters cannot be turned on (err %d)\n", err);
+	}
+}
+
+
+static void dfu_smp_c_on_error(struct bt_gatt_dfu_smp_c *dfu_smp_c, int err)
+{
+	printk("DFU SMP generic error: %d\n", err);
+}
+
+
+static const struct bt_gatt_dfu_smp_c_init_params init_params = {
+	.error_cb = dfu_smp_c_on_error
+};
+
+static CborError cbor_stream(void *token, const char *fmt, ...)
+{
+	va_list ap;
+
+	(void)token;
+	va_start(ap, fmt);
+	vprintk(fmt, ap);
+	va_end(ap);
+
+	return CborNoError;
+}
+
+
+static void smp_echo_rsp_proc(struct bt_gatt_dfu_smp_c *dfu_smp_c)
+{
+	u8_t *p_outdata = (u8_t *)(&smp_rsp_buff);
+	const struct bt_gatt_dfu_smp_rsp_state *rsp_state;
+
+	rsp_state = bt_gatt_dfu_smp_c_rsp_state(dfu_smp_c);
+	printk("Echo response part received, size: %u.\n",
+	       rsp_state->chunk_size);
+
+	if (rsp_state->offset + rsp_state->chunk_size > sizeof(smp_rsp_buff)) {
```

Comment:
any chance for an overflow on `rsp_state->offset + rsp_state->chunk_size` e.g. when offset is huge?

Reply:
Technically chunk size is limited to MTU - I doubt it can roll over as the offset would be outside of RAM much earlier.